salt/modules/transactional_update.py

```diff
@@ -985,13 +985,30 @@ def call(function, *args, **kwargs):
             else:
                 return local
         except ValueError:
-            return {"result": False, "retcode": 1, "comment": ret_stdout}
+            local = {"result": False, "retcode": 1, "comment": ret_stdout}
+            return local
     finally:
         # Check if reboot is needed
-        if activate_transaction and pending_transaction():
+        if (activate_transaction and pending_transaction()) or _user_specified_reboot(
+            local
+        ):
             reboot()
 
 
+def _user_specified_reboot(local):
+    if not isinstance(local, dict):
+        # Skip if execution is not state/highstate
+        return False
```

Comment:
This does skip if this is not a state application, but it will also match arbitrary execution modules that return dicts.